File: backend/app/services/notification_service.py
"""Notification creation + realtime fan-out."""
from __future__ import annotations

from app.models.notification import Notification
from app.repositories.misc_repo import NotificationRepository
from app.websockets.manager import connection_manager


class NotificationService:
    def __init__(self, notifications: NotificationRepository) -> None:
        self.notifications = notifications

    async def notify(
        self, user_id: str, *, type_: str, title: str, body: str | None = None,
        link: str | None = None,
    ) -> Notification:
        notif = await self.notifications.insert(
            Notification(user_id=user_id, type=type_, title=title, body=body, link=link)
        )
        await connection_manager.send_to_user(
            user_id,
            {"event": "notification.new", "data": notif.model_dump(mode="json")},
        )
        return notif

    async def list(self, user_id: str) -> dict:
        items = await self.notifications.list_for_user(user_id)
        unread = await self.notifications.unread_count(user_id)
        return {
            "items": [n.model_dump(mode="json") for n in items],
            "unread": unread,
        }

    async def mark_read(self, user_id: str, ids: list[str] | None = None) -> None:
        await self.notifications.mark_read(user_id, ids)

File: backend/src/services/notificationService.js
"use strict";

/** Notification creation + realtime fan-out. Mirrors app/services/notification_service.py. */

const { Notification } = require("../models");
const mongoose = require("mongoose");
const connectionManager = require("../sockets/manager");

async function notify(userId, { type, title, body = null, link = null }) {
  const notif = await Notification.create({
    user_id: userId,
    type,
    title,
    body,
    link,
  });
  await connectionManager.sendToUser(userId, {
    event: "notification.new",
    data: notif.toJSON(),
  });
  return notif;
}

async function list(userId) {
  const items = await Notification.find({ user_id: userId })
    .sort({ created_at: -1 })
    .limit(30);
  const unread = await Notification.countDocuments({ user_id: userId, read: false });
  return {
    items: items.map((n) => n.toJSON()),
    unread,
  };
}

async function markRead(userId, ids = null) {
  const query = { user_id: userId, read: false };
  if (ids && ids.length) {
    const validIds = ids.filter((i) => mongoose.isValidObjectId(i));
    query._id = { $in: validIds };
  }
  await Notification.updateMany(query, { $set: { read: true } });
}

module.exports = { notify, list, markRead };
